google/auth/_credentials_base.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Interface for base credentials."""
+
+import abc
+from enum import Enum
+
+from google.auth import _helpers
+from google.auth import metrics
+from google.auth._refresh_worker import RefreshThreadManager
+
+_DEFAULT_UNIVERSE_DOMAIN = "googleapis.com"
+
+
+class _BaseCredentials(metaclass=abc.ABCMeta):
```

Comment:
I am not convinced a shared base class between sync and async credentials is necessary.  I do not think the code is large enough / complex enough that the tradeoff between reduced code size and an additional inheritance is worthwhile.

Do you mind sharing some of the benefits?

Reply:
I agree that introducing a base class here is not necessary. But it will prevent code duplication for when we implement the async class.

Avoiding repetition is the main benefit. If this is a sticking point we can omit the base class for now so it doesn't block the Async REST project, and it can be added later. Let me know what you think!

Reply:
I would argue for having the base class, even if small. It's an investment in simplicity and lowering future maintenance by having shared logic in one place. Not having had this done in the past in other code has become a source of friction for us subsequently working on that code.

While there is a trivial performance cost in using inheritance [O(µs)], IIUC this code would not be used in any tight inner loops where it would be noticeable by the user. (Correct me if I'm wrong, please.)